File: 11_leetcode/01_easy/00009/easy9.go
package main

import (
	"fmt"
	"strconv"
)

func main() {
	fmt.Println(isPalindrome(121))
	fmt.Println(isPalindrome(1221))
	fmt.Println(isPalindrome(-121))
	fmt.Println(isPalindrome(10))
	fmt.Println(isPalindrome(1))
	fmt.Println()
	fmt.Println(isPalindrome2(121))
	fmt.Println(isPalindrome2(1221))
	fmt.Println(isPalindrome2(-121))
	fmt.Println(isPalindrome2(10))
	fmt.Println(isPalindrome2(1))
}

// 実行速度こっちのほうが実行速度が早い
func isPalindrome(x int) bool {
	strX := strconv.Itoa(x)
	n := len(strX)
	for i := 0; i < n/2; i++ {
		// 逆から見ていく
		if strX[i] != strX[n-1-i] {
			// 違う時点で回分ではない
			return false
		}
	}
	return true
}

func isPalindrome2(x int) bool {
	strX := strconv.Itoa(x)

	left, right := 0, len(strX)-1
	for left < right {
		if strX[left] != strX[right] {
			return false
		}
		left++
		right--
	}

	return true
}

File: 11_leetcode/01_easy/00009/Easy9.java
public class Easy9 {
    public static void main(String[] args) {
        Easy9 ins = new Easy9();
        System.out.println(ins.isPalindrome(121));
        System.out.println(ins.isPalindrome(1221));
        System.out.println(ins.isPalindrome(-121));
        System.out.println(ins.isPalindrome(10));
        System.out.println(ins.isPalindrome(1));
        System.out.println("aaa");
        System.out.println(ins.isPalindrome2(121));
        System.out.println(ins.isPalindrome2(1221));
        System.out.println(ins.isPalindrome2(-121));
        System.out.println(ins.isPalindrome2(10));
        System.out.println(ins.isPalindrome2(1));
    }

    public boolean isPalindrome(int x) {
        if(x<0){
            return false;
        }

        String str = Integer.toString(x);
        int length = str.length();

        for (int i=0;i<length/2;i++){
            // パースする必要はない
            // int start = Integer.parseInt(str.substring(i,i+1));
            // int end = Integer.parseInt(str.substring(length-i-1,length-i));
            char start = str.charAt(i);
            char end = str.charAt(length-i-1);
            if (start != end){
                return false;
            }
        }
        return true;
    }

    public boolean isPalindrome2(int x) {

        String strX = Integer.toString(x);

        int left = 0;
        int right = strX.length() - 1;

        while (left < right) {
            if (strX.charAt(left) != strX.charAt(right)) {
                return false;
            }
            left++;
            right--;
        }
        return true;
    }
}
